perf/collector_libpfm.go

```diff
@@ -0,0 +1,250 @@
+package perf
+
+// #cgo CFLAGS: -I/usr/include
+// #cgo LDFLAGS: -lpfm
+// #include <perfmon/pfmlib.h>
+// #include <stdlib.h>
+import "C"
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"sync"
+	"unsafe"
+
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/stats"
+
+	"golang.org/x/sys/unix"
+	"k8s.io/klog"
+)
+
+type collector struct {
+	cgroupPath   string
+	events       Events
+	cpuFiles     map[string]map[int]readerCloser
+	cpuFilesLock sync.Mutex
+	numCores     int
+}
+
+var (
+	isLibpfmInitialized = false
+	libpmfMutex         = sync.Mutex{}
+)
+
+func init() {
+	libpmfMutex.Lock()
+	defer libpmfMutex.Unlock()
+	pErr := C.pfm_initialize()
+	if pErr != C.PFM_SUCCESS {
+		fmt.Printf("unable to initialize libpfm: %d", int(pErr))
+		return
+	}
+	isLibpfmInitialized = true
+}
+
+func NewCollector(cgroupPath string, events Events, numCores int) stats.Collector {
+	return &collector{cgroupPath: cgroupPath, events: events, cpuFiles: map[string]map[int]readerCloser{}, numCores: numCores}
+}
+
+func (c *collector) UpdateStats(stats *info.ContainerStats) error {
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+
+	stats.PerfStats = []info.PerfStat{}
+	klog.V(4).Infof("Attempting to update perf_event stats from cgroup %q", c.cgroupPath)
+	for name, files := range c.cpuFiles {
+		for cpu, file := range files {
+			buf := make([]byte, 32)
+			_, err := file.Read(buf)
+			if err != nil {
+				klog.Warningf("Unable to read from perf_event file (event: %q, CPU: %d) for %q", name, cpu, c.cgroupPath)
+				continue
+			}
+			perfData := &ReadFormat{}
+			reader := bytes.NewReader(buf)
+			err = binary.Read(reader, binary.LittleEndian, perfData)
+			if err != nil {
+				klog.Warningf("Unable to decode from binary format read from perf_event file (event: %q, CPU: %d) for %q", name, cpu, c.cgroupPath)
+			}
+			klog.V(4).Infof("Read metric for event %q for cpu %d from cgroup %q: %d", name, cpu, c.cgroupPath, perfData.Value)
+			scalingRatio := 1.0
+			if perfData.TimeEnabled != 0 {
+				scalingRatio = float64(perfData.TimeRunning) / float64(perfData.TimeEnabled)
+			}
+			stat := info.PerfStat{
+				Value:        uint64(float64(perfData.Value) / scalingRatio),
+				Name:         name,
+				ScalingRatio: scalingRatio,
+				Cpu:          cpu,
+			}
+			stats.PerfStats = append(stats.PerfStats, stat)
+		}
+	}
+	return nil
+}
+
+func (c *collector) Setup() {
+	cgroup, err := os.Open(c.cgroupPath)
+	if err != nil {
+		klog.Warningf("Unable to open cgroup directory %s: %s", c.cgroupPath, err)
+	}
+	defer cgroup.Close()
+
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+	cgroupFd := int(cgroup.Fd())
+	for _, group := range c.events.Events {
+		if len(group) == 1 {
+			customEvent, ok := c.events.eventToCustomEvent[group[0]]
+			if ok {
+				c.setupRawNonGrouped(customEvent, cgroupFd)
+			} else {
+				c.setupNonGrouped(string(group[0]), cgroupFd)
+			}
+		} else {
+			klog.Info("Grouped events are not supported yet")
```

Comment:
Can we validate this during startup?  Is this something you are planning to implement before we submit the change?

Reply:
I was originally thinking about ignoring such setting but I now I think that returning noop manager early might be much better idea.